Adaptive average pooling 3d

ssnl opened the pull request:
Adding AdaptiveAvgPool3d as per request in #1988 .

Test plan:
./run_test.sh

SsnL commented:
Hi there, this is my first time contributing. Let me know if I did anything wrong/suboptimally. :)

soumith commented:
btw before you start on AdaptiveMaxPool3d, see the changes Greg made for the 2d class today: https://github.com/pytorch/pytorch/pull/2722

soumith commented:
@pytorchbot add to whitelist

gchanan commented on test/test_nn.py:
```diff
@@ -4024,6 +4024,24 @@ def add_test(test):
     ),
     dict(
+        module_name='AdaptiveAvgPool3d',
+        constructor_args=((3),),
```
nit: the extra parens don't actually do anything here?

gchanan commented on torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c:
```diff
@@ -0,0 +1,302 @@
+          real *input_p,
+          real *output_p,
+          long nslices,
+          long isizeT,
```
can you just use the same variable names / name formats as in SpatialAdaptiveAveragePooling?  Even if your names are better (they are), this creates unnecessary noise in review and maintenance.

SsnL replied:
I debated against myself on that for some time too. But I wanted to be consistent between the cpu and the cuda var names. I really didn't want to follow the spatial [cuda var names](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/SpatialAdaptiveAveragePooling.cu#L21) which leads to `input_t` being a var rather than a type. What's your though on this?

soumith replied:
if that's the case, just codemod the names in Spatial then to isize*.

SsnL replied:
Should I do that in this PR or a separate one?

soumith replied:
doing in this PR is fine.

SsnL replied:
Will do tomorrow.

gchanan commented on torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c:
```diff
@@ -0,0 +1,302 @@
+  {
+    /* loop over output */
+    long ti, i, j;
+    for(ti = 0; ti < osizeT; ti++)
```
give "ti" a different name; from the name I'd expect it to loop over the time dimension of the input ("ot" doesn't seem unreasonable.

gchanan commented on torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c:
```diff
@@ -0,0 +1,302 @@
+
+    /* calculate average */
+    long ti, i, j;
+    for(ti = 0; ti < osizeT; ti++)
```
same here (both "isizeT", etc. and "ti"

gchanan reviewed (commented):
I haven't looked at the cuda code yet (I'll do it tomorrow), but so far looks good!  Comments are minaly nits around minimizing the differences in the code with the Spatial versions.

SsnL commented on torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c:
```diff
@@ -0,0 +1,302 @@
+              for(x = x_start; x < x_end; x++)
+              {
+                /* update gradient */
+                gradInput_p_k[z*isizeH*isizeW + y*isizeW + x] += grad_delta;
```
I'm suddenly confused by this line (and the corresponding line in spatial version). Why is this not atomic?

soumith replied:
dont rule out bugs :)
It does indeed look like a race

ruotianluo replied:
Because it's parallelized batch and channel dimension.

SsnL replied:
Thanks for enlightening me, @ruotianluo !

SsnL commented:
Did massive renaming. I'd appreciate if someone can take a look. :)

gchanan commented on torch/lib/THCUNN/generic/SpatialAdaptiveAveragePooling.cu:
```diff
@@ -20,59 +22,57 @@ void THNN_(SpatialAdaptiveAveragePooling_updateOutput)(
+    int isizeH = input->size[1];
+    int isizeW = input->size[2];
 
-    long istride_d = input->stride[0];
```
why change these from long to int?

gchanan commented on torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c:
```diff
@@ -7,49 +7,51 @@
           real *output_p,
-          long nslices,
-          long iwidth,
-          long iheight,
```
same thing, why did these change?

gchanan reviewed (commented):
1) Your changes now conflict, can you rebase and fix?
2) The spatial changes look correct, although I don't know why you changed some long types to int.  Can you explain? (I assume you meant to make them consistent, but don't know why you preferred int to long).
3) In general, it would be helpful if you made this easier to review.  For example, you could dedicate commits to just the spatial renames (no logic changes), then include logic changes in the next commit, then adding Volumetric files that are as close to the (changed) Spatial versions as possible, with clear commit messages about what each does.  I'm not asking you to change these now (except the above), just a note for the future.

SsnL commented:
Thanks for your input @gchanan .

1 & 2. I preferred int because (a) in CUDA version, the kernel using them has int input parameters and (b) in CPU & CUDA version, the input indices computed from START_IND and END_IND are set to type int. So using long didn't seem necessary to me. Should I stick to using long (now int64_t in the master)?
The int64_t's are causing conflicts.  I will rebase & merge once this is settled.
3. Thanks for the tip! I will keep this in mind.

gchanan commented:
> I preferred int because (a) in CUDA version, the kernel using them has int input parameters and (b) in CPU & CUDA version, the input indices computed from START_IND and END_IND are set to type int. So using long didn't seem necessary to me. Should I stick to using long (now int64_t in the master)?

Good question.  I discussed briefly with a couple of people and my best guess is this:
- In general, we only really care about performance on CUDA.  So the correct and performant thing to do there is to operate on ints if we can, otherwise operate on longs (you see this by searching for `canUse32BitIndexMath`.  Some of the implementations are lazy about this (or use some knowledge about the grid size) and always use ints.
- On cpu, we usually don't care that much and usually just use `long`.
- Some of your changes look incorrect.  For example, storing the strides as `int`s which are then applied as pointer arithmetic would previously work if everything but strides fit into `int`, but now they are broken.

In short, without a clear motivation beyond making everything the same type, I'm loathe to change the types.

SsnL commented:
Thanks @gchanan . I'm working on an update right now. I pushed some commits because I need to switch to another develop machine. I will notify here when the PR is ready to be reviewed.

SsnL commented:
Hi @gchanan , I've addressed your comments in new commits. Now:
1. The `int64_t`s (`long`s) in spatial versions stay as `int64_t`s.
2. Corresponding variables in volumetric versions are now `int64_t`s.
3 I propose changing spatial CUDA kernel's `stride` inputs to type `int64_t` because the caller has them in `int64_t`. So this fixes the case that everything but strides fit into `int`s in CUDA spatial case. This is done in a separate commit `changed stride in spatial CUDA kernel to int64_t`.
4. Rebased & merged.

Thanks again for your comments!

gchanan reviewed (changes requested):
The python/docs changes all look good.  I'm not planning to look at those again unless you tell me they changed.

I'm still not really confident I can review the C/CUDA changes properly as is; the first commit contains variable renames, ordering changes, and logic changes it ways that don't seem consistent (i.e. I think I caught some mistakes but it's hard to tell with the mass of changes and it's likely I missed some others).  There also seems to be an effort to maintain consistency across C/CUDA implementations, but not across Spatial/Volumetric with types and such; that doesn't seem like the right axis to maintain consistency on; they are different languages, you are optimizing for different things, etc.

Here is how I think an "ideal" review of the C/CUDA would go:
1) First commit(s) contain existing Spatial code with only renames
2) Next commit(s) contain logic/reordering changes to existing Spatial code
3) Next commit(s) contain Volumetric changes.  Volumetric changes are verified by diffing the corresponding Spatial versions (after 2 above) against the Volumetric versions.  The only differences are due to it being Volumetric, i.e. extra for loops, stride calculations.

Does that make sense?

SsnL commented on torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c:
```diff
@@ -7,49 +7,51 @@
     /* loop over output */
-    int64_t i, j;
-    for(i = 0; i < oheight; i++)
+    int oh, ow;
```
Missed that. Sorry.

ssnl committed:
rename spatial version

ssnl committed:
reorder spatial version variables so that B (batch) before D (feature) before H (height) before W (width); change some code to be more concise

SsnL commented:
Thanks @gchanan ! I've restructured the commits and `push --force`. Sorry for causing these inconvenience. In my other PR (adaptive max pool 3d), I tried to follow your advice and structure them better. So that one should be much easier to review :). 

Let me know if there is anything more I need to change!

SsnL commented on torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c:
```diff
@@ -89,96 +92,99 @@ void THNN_(SpatialAdaptiveAveragePooling_updateOutput)(
-                                                        istride_d);
+      THNN_(SpatialAdaptiveAveragePooling_updateOutput_frame)(input_data+b*istrideB, output_data+b*sizeD*osizeH*osizeW,
+                                                        sizeD,
+                                                        isizeH, isizeW,
```
Thanks for pointing that out. Resolved in updated commits!

gchanan commented:
btw, closing and reopening the PR should trigger travis so you don't have to do an empty commit.

SsnL commented:
Thanks for the tip @gchanan :)

gchanan commented on torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu:
```diff
@@ -0,0 +1,173 @@
+  // each H*W plane is processed by blocksH thread blocks
+  int blocksH = max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
```
why did you choose to break things up this way (blocks,threads) (where the blocks are 2d), instead of the existing Volumetric pooling (VolumetricAveragePooling, VolumetricDilatedMaxPooling) where it's (grid/blocks) and grid is 3d?

SsnL replied:
`(blocks,threads)` naming is following the SpatialAdaptiveAveragePooling.
That blocks is 2d is also following SpatialAdaptiveAveragePooling, where each thread does multiple pixels rather than 1 (`blocksH` blocks together process a H*W plane). In VolumetricDilatedMaxPooling, H*W/(32*8) blocks process a plane. 

I'm not sure if one way is better than the other. But I thought following spatial version should be a reasonable thing to do. :)

gchanan reviewed (commented):
Looks good.  Non-CUDA changes are ready; I haven't done much CUDA programming, so I'm less confident about that.  If you could explain my question below it would help.

SsnL commented on torch/lib/THCUNN/VolumetricAdaptiveAveragePooling.cu:
```diff
@@ -0,0 +1,250 @@
+#include "THCUNN.h"
+#include "THCHalf.h"
+#include "THCDeviceTensor.cuh"
```
Good catch! They are not needed. I forgot to remove them after removing a function call. I updated the PR.

ssnl committed:
Spatial CUDA kernel: removed unused sizeD parameter; changed stride types to int64_t to be consistent with caller function

ssnl committed:
VolumetricAdaptiveAveragePool

gchanan commented on torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu:
```diff
@@ -0,0 +1,173 @@
+    osizeW = gradOutput->size[4];
+  }
+
+  // somehow nonatomic is passing all test for volumetric case.
```
it should, assuming it is written correctly, right?  My understanding is that nonatomic means it's not using any atomic operations (e.g. atomicAdd); since it's operating over the input pixels this is okay because only one thread is writing to that location in the grad input.  The atomic version is operating over the output pixels; since an input can correspond to multiple outputs it needs to use atomic ops.

SsnL replied:
Yeah, it should. Spatial version said its nonatomic kernel isn't passing the tests and thus disabled it. Either there is something weird going on with spatial kernel, or the author just copied from spatial adaptive max pooling, which also has that line (https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/generic/SpatialAdaptiveMaxPooling.cu#L97). But my volumetric adaptive max pooling also works for the nonatomic case in tests #2782 , so I'm not really sure what's going on.

fmassa replied:
For the record, if I remember properly the `SpatialAdaptiveMaxPooling` didn't indeed pass the tests without atomic operations , so I just used it by default and didn't look back to see why it was the case :-)

SsnL replied:
Thanks for clearing things up :)

gchanan reviewed (approved):
looks good!

soumith commented:
thanks a ton @SsnL !!!

SsnL commented:
Thanks @soumith ! Glad to contribute!
